idx/memory/meta_tags_test.go

```diff
@@ -0,0 +1,268 @@
+		So(err, ShouldBeNil)
+		So(record, ShouldNotBeNil)
+		So(oldHash, ShouldBeZeroValue)
+		So(oldRecord, ShouldBeNil)
+
+		Convey("then it should exist in the meta tag records object", func() {
+			So(len(metaTagRecords), ShouldEqual, 1)
+			record, ok := metaTagRecords[record.hashQueries()]
+			So(ok, ShouldBeTrue)
+
+			So(len(record.metaTags), ShouldEqual, 2)
+			So(len(record.queries), ShouldEqual, 2)
+
+			var seenMetaTag1, seenMetaTag2 bool
+			for _, metaTag := range record.metaTags {
+				if reflect.DeepEqual(metaTag, kv{key: "metaTag1", value: "abc"}) {
+					seenMetaTag1 = true
+				}
+				if reflect.DeepEqual(metaTag, kv{key: "anotherTag", value: "theValue"}) {
+					seenMetaTag2 = true
+				}
+			}
+			So(seenMetaTag1, ShouldBeTrue)
+			So(seenMetaTag2, ShouldBeTrue)
+
+			var seenQuery1, seenQuery2 bool
+			for _, query := range record.queries {
+				if reflect.DeepEqual(query, expression{kv: kv{key: "metricTag", value: "a"}, operator: NOT_EQUAL}) {
+					seenQuery1 = true
+				}
+				if reflect.DeepEqual(query, expression{kv: kv{key: "match", value: "this"}, operator: MATCH}) {
+					seenQuery2 = true
+				}
+			}
+			So(seenQuery1, ShouldBeTrue)
+			So(seenQuery2, ShouldBeTrue)
+		})
+	})
+}
+
+func TestUpdateExistingMetaTagRecord(t *testing.T) {
+	// define the values for two metric records
+	metaTags1 := []string{"metaTag1=value1"}
+	tagQueries1 := []string{"tag1=~a", "tag2=~b"}
+	metaTags2 := []string{"metaTag1=value1"}
+	tagQueries2 := []string{"tag1=~c", "tag2=~d"}
+
+	// define the values for an update which is going to replace
+	// the first metric record because it has the same tag queries
+	metaTagsUpdate := []string{"metaTag1=value2"}
+	tagQueriesUpdate := []string{"tag1=~a", "tag2=~b"}
+
+	metaTagRecords := make(metaTagRecords)
+
+	Convey("When adding two meta tag records", t, func() {
+		_, record, oldHash, oldRecord, err := metaTagRecords.upsert(metaTags1, tagQueries1)
+		So(err, ShouldBeNil)
+		So(record, ShouldNotBeNil)
+		So(oldHash, ShouldBeZeroValue)
+		So(oldRecord, ShouldBeNil)
+
+		_, record, oldHash, oldRecord, err = metaTagRecords.upsert(metaTags2, tagQueries2)
+		So(err, ShouldBeNil)
+		So(record, ShouldNotBeNil)
+		So(oldHash, ShouldBeZeroValue)
+		So(oldRecord, ShouldBeNil)
+
+		Convey("then they should exist", func() {
+			So(len(metaTagRecords), ShouldEqual, 2)
+
+			// the order of the records may have changed due to sorting by hash
+			var record1, record2 metaTagRecord
+			var found1, found2 bool
+			var recordIdToUpdate uint32
+			for hash, record := range metaTagRecords {
+				So(len(record.queries), ShouldEqual, 2)
+				switch record.queries[0].value {
+				case "a":
+					record1 = metaTagRecords[hash]
+					found1 = true
+					recordIdToUpdate = hash
+				case "c":
+					record2 = metaTagRecords[hash]
+					found2 = true
+				}
+			}
+
+			// verify that we found both records
+			So(found1 && found2, ShouldBeTrue)
+
+			So(record1.metaTags[0], ShouldResemble, kv{key: "metaTag1", value: "value1"})
```

Comment:
At the time I wrote that I thought exactly the same, because the word "resemble" sounds like it's not strict. It does "deep equal" though, which is actually stricter than the word "resemble" makes it seem: https://github.com/smartystreets/goconvey/wiki/Assertions#general-equality

Reply:
i'll add it to the list of reasons why i dislike convey :]

Reply:
lol, didn't know that you don't like convey. I can stop using it if you don't like it, makes no difference to me. the only good thing that I see in convey is this whole web interface that comes with it, which shows you the test stats etc.